### ISSUE TITLE
Provide public API to generate data based on factory

### ISSUE THREAD

**Issue by jelhan** [samselikoff/ember-cli-mirage#1581]
Having the ability to generate dummy data based on factories is helpful for acceptance testing of create and edit functionality. This is currently only possible using not documented private API `server.build('type')`. It would be awesome if we could expose this API publicly. The return should be the same as `sever.create()` but without persisting the data. `build()` is not the best name for such a method in my opinion. I would expect it to be named `make()`. I don't think there is a need for a `makeList()` method.

Usage example:

```js
test('create a post', async function(assert) {
  let data = this.server.make('post');
  
  await vist('/create-post');
  await fillIn('[data-test-form-element-for="title"] input', data.title);
  await fillIn('[data-test-form-element-for="text"] textarea', data.text);
  await click('[data-test-button="submit"]');

  assert.deepEqual(data, this.server.db.posts[0], 'post has been persisted with correct data');
});
```

**Comment by samselikoff** [comment]
Hmm... The problem I see with this test is that it's using a server factory to fill in input values on a client app. If we think about how best to approximate the real-world behavior of this app, the user would type in values on their own. I'm just not sure Mirage's factories should hold this type of information.

Now I'd like to see more of your tests, because I've personally never had this pain point. I might write the test like this:

```js
test('create a post', async function(assert) {
  let originalLength = this.server.db.posts.length;
  
  await vist('/create-post');
  await fillIn('[data-test-form-element-for="title"] input', "Post title");
  await fillIn('[data-test-form-element-for="text"] textarea', "Post body");
  await click('[data-test-button="submit"]');

  let newLength = this.server.db.posts.length;
  let newPost = this.server.db.posts[this.server.db.posts.length];
  
  assert.ok(newLength, originalLength + 1, 'a new post was created');
  assert.deepEqual(newPost.title, "Post title", 'post has been persisted with correct title');
  assert.deepEqual(newPost.body, "Post body", 'post has been persisted with correct body');
});
```

If the attribute assertion were too much (and honestly I usually don't assert against all the attributes, usually just the `id`), I would wrap up the input values in an object:

```js
test('create a post', async function(assert) {
  let originalLength = this.server.db.posts.length;
  let newData = { title: 'Post title', body: 'Post body' };
  
  await vist('/create-post');
  await fillIn('[data-test-form-element-for="title"] input', newData.title);
  await fillIn('[data-test-form-element-for="text"] textarea', newData.body);
  await click('[data-test-button="submit"]');

  let newLength = this.server.db.posts.length;
  let newPost = this.server.db.posts[this.server.db.posts.length];
  
  assert.ok(newLength, originalLength + 1, 'a new post was created');
  assert.deepEqual(newPost, newData, 'post has been persisted with correct attrs');
});
```

but honestly I don't write tests like this much. I don't really like abstracting much in tests, I think it's way too easy to get carried away, make tests brittle, or test the wrong things. And in any case, the kinds of things a Mirage server factory would generate might overlap with but are not necessarily the same thing as the kinds of things you'd want in the `newData` object.

Just my initial thoughts.

**Comment by jelhan** [comment]
Thanks a lot for your detailed answer. I'm sorry for not replying back earlier.

I think the main question here is if factories should be used in testing at all. If I got you right, factories are adding to much abstraction to tests and therefore might make the tests less useful. This opinion does not only apply to factories being used to test create or edit functionality but also for showing data.

Let me give a test using factories and not using factories as short examples of what I mean:

```js
// not using factories
test("I can view the movie title", async function(assert) {
  let movie = this.server.create('movie', { title: "Interstellar" });

  await visit(`/movies/${movie.id}`);

  assert.dom('h1').includesText("Interstellar");
});

// using factories
test("I can view the movie title", async function(assert) {
  let movie = this.server.create('movie');

  await visit(`/movies/${movie.id}`);

  assert.dom('h1').includesText(move.title);
});
```

This all comes done to the question if one should use random input data in tests or not. I'm not sure if Mirage should have an opinion about random testing. It should support both as first class citizens in my opinion. Currently it makes random testing unnecessary difficult by restricting factories to persisted records only.

**Comment by samselikoff** [comment]
Good point. Even though I do like option 1 better, I think we've been too restrictive & high level in the past and I wish we had built more lower-level primitives into Mirage and allowed the higher-level patterns to emerge.

I will say I don't think it's all-or-nothing. I like option 1 here because the title is relevant to the assertion; but I think it's fine and even desirable for factories to abstract details that aren't relevant to the test from the actual test code. This is one of their strengths and how they make test suites less brittle.

I think I would be ok exposing `make` as public API, and just adding some notes to the docs about how I think about some of these issues 👍

**Comment by elwayman02** [comment]
I'd like to throw in my two cents that it would be super useful to document using factories directly, for use-cases like integration tests that would consume the same mocked data but don't fire any network requests of their own. I'd like to be able to do something like this:

```
import FooFighterFactory from 'dummy/mirage/factories/foo-fighter';

...

hooks.beforeEach(function () {
  this.fooFighterFactory = new FooFighterFactory();
});

...
test('some foo test', async function (assert) {
  this.fooFighter = this.fooFighterFactory.build('withExtraFooTrait');
  
  await render(hbs`{{foo-fighter foo=fooFighter}}`);
});
```

I know I don't have the syntax for how the factory `build` method works exactly right, because it's not documented. I definitely want to figure out how to do this, though!

**Comment by jelhan** [comment]
@elwayman02 The syntax of `server.build()` is basically the same as the one of `server.create()`, which uses `server.build()` under the hood: https://github.com/samselikoff/ember-cli-mirage/blob/cf90e92842ea5cff23f2a71093af7e3cee89f75f/addon/server.js#L691

Please note that both require an instance of mirage server. So you have to start the server nevertheless. Importing a factory and using that one directly is not supported yet as far as I'm familiar with code base.
